### DISCUSSION TITLE
middleware: assert message["type"] == "http.response.start"

### DISCUSSION THREAD

**Discussion by lazka** [fastapi/fastapi#7431] [Questions] [answered]
### Describe the bug

I've tried to add a middleware to my app and then I started hitting a assertation error with the test client.

### To Reproduce

Smallest reproducer I could find. It needs jinja2+middlware+testclient to hit the error.

```python
import os
import tempfile
from typing import Callable
from fastapi import FastAPI

from fastapi import Request, Response
from fastapi.testclient import TestClient
from fastapi.templating import Jinja2Templates

dir_ = tempfile.mkdtemp()
with open(os.path.join(dir_, "index.html"), "w"):
    pass

app = FastAPI()
templates = Jinja2Templates(directory=dir_)


@app.get("/")
async def index(request: Request) -> Response:
    return templates.TemplateResponse("index.html", {"request": request})


@app.middleware("http")
async def some_middleware(request: Request, call_next: Callable) -> Response:
    response: Response = await call_next(request)
    return response


with TestClient(app) as client:
    client.get("/")
```

Error (partial)

```python
  File "/home/lazka/.cache/pypoetry/virtualenvs/msys2-web-tQnlPJcR-py3.8/lib/python3.8/site-packages/starlette/middleware/errors.py", line 159, in __call__
    await self.app(scope, receive, _send)
  File "/home/lazka/.cache/pypoetry/virtualenvs/msys2-web-tQnlPJcR-py3.8/lib/python3.8/site-packages/starlette/middleware/base.py", line 25, in __call__
    response = await self.dispatch_func(request, self.call_next)
  File "a.py", line 25, in some_middleware
    response: Response = await call_next(request)
  File "/home/lazka/.cache/pypoetry/virtualenvs/msys2-web-tQnlPJcR-py3.8/lib/python3.8/site-packages/starlette/middleware/base.py", line 47, in call_next
    assert message["type"] == "http.response.start"
AssertionError
```

### Expected behavior

Adding a no-op middleware shouldn't make the TestClient error out.

**Comment by lazka** [comment]
Still an issue. I'm wondering, is no one using a middleware + running tests? Or am I missing something maybe?

**Comment by Kludex** [comment, marked as answer]
Known issue: https://github.com/encode/starlette/issues/472

Solution: https://github.com/tiangolo/fastapi/issues/806 :tada: 

EDIT (December 17th, 2022): I do not recommend that solution anymore.
EDIT (April 3rd, 2023): The issue was closed on Starlette's side with a solution. :)

**Comment by lazka** [comment]
Thanks, good to know that it's not just me.

**Comment by tiangolo** [comment]
It seems this would be an issue in Starlette and not here, right? In that case, it would make sense to close this one and keep the one there. 🤓

**Comment by lazka** [comment]
Fine with me, though the downside is that other users might not find this issue then.

**Comment by tiangolo** [comment]
Thanks! The issue would not be removed, only closed. So if they for that they will find it, just closed, and it will help me (try to) keep things a bit cleaner.

Given that, I'll then close it. Thanks!
